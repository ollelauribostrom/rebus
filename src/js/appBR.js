import { clickReset } from './components/ResetButton';

export function loadButtons(resetIdiom) {
  clickReset(resetIdiom);
  clickbr();
  clicken();
}

// BUTTON [PT-BR] - LOAD BUTTON
export function clickbr() {
  // CONDITION: if init as [pt-br] rebus, change design and hide button [ptbr]
  // ELSE: hide button [english]
  checkIsPTBR();

  const e = document.getElementById('button-ptbr'); // get button [pt-br]

  // HOVER FUNCTION: cursor (hover) [pt-br]
  mouseHoverPTBR(e);

  // CLICK FUNCTION: click on button [pt-br]
  e.addEventListener('click', () => {
    localStorage.setItem('flagBR', 'true');
  });
  // FLAG: now [pt-br] rebus
}

// BUTTON [EN] - LOAD BUTTON
export function clicken() {
  const a = document.getElementById('button-en'); // get button [english]

  // HOVER FUNCTION: cursor (hover) [english]
  mouseHoverEN(a);

  // CLICK FUNCTION: click on button [english]
  a.addEventListener('click', () => {
    localStorage.setItem('flagBR', 'false');
  });
  // FLAG: now [english / original] rebus
}

function mouseHoverPTBR(e) {
  e.addEventListener('mouseover', () => {
    document.getElementById('button-text-id-br').classList.add('button-text-show');
  });
  e.addEventListener('mouseout', () => {
    document.getElementById('button-text-id-br').classList.remove('button-text-show');
  });
}

function mouseHoverEN(a) {
  a.addEventListener('mouseover', () => {
    document.getElementById('button-text-id-en').classList.add('button-text-show');
  });
  a.addEventListener('mouseout', () => {
    document.getElementById('button-text-id-en').classList.remove('button-text-show');
  });
}

export function checkResetPTBR() {
  let resetIdiom = 'reset-english';
  if (localStorage.getItem('flagBR') === 'true') resetIdiom = 'reset-ptbr';
  return resetIdiom;
}

export function checkIsPTBR() {
  if (localStorage.getItem('flagBR') === 'true') {
    return (
      document.querySelector('.root').classList.add('rootbr'),
      document.getElementById('button-ptbr').classList.add('button-spam')
    );
  }
  return document.getElementById('button-en').classList.add('button-spam');
}
