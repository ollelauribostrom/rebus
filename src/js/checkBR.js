export function checkResetPTBR(param) {
  let resetIdiom = param;
  if (localStorage.getItem('flagBR') === 'true') {
    resetIdiom = 'reset-ptbr';
  } else {
    resetIdiom = 'reset-english';
  }
  return resetIdiom;
}

export function checkIsPTBR() {
  if (localStorage.getItem('flagBR') === 'true') {
    document.querySelector('.root').classList.add('rootbr');
    document.getElementById('button-ptbr').classList.add('button-spam');
  } else {
    document.getElementById('button-en').classList.add('button-spam');
  }
}
