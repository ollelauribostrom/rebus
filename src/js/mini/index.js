export { createComponent } from './component';
export { render } from './render';
export { createStore } from './store';
