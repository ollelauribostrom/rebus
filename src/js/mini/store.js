export function createStore(initialState, actions) {
  const store = {
    connectedComponents: [],
    state: initialState
  };
  store.actions = new Proxy(actions, {
    get(target, key) {
      return async (...args) => {
        const newState = await actions[key](store.state, ...args);
        store.state = Object.assign(store.state, newState);
        store.connectedComponents.forEach(c => c.update());
      };
    }
  });
  store.connect = component => {
    const connected = new Proxy(component, {
      get(target, key) {
        if (key === 'render') {
          return (...args) => target.render(store.state, ...args);
        }
        return Reflect.get(target, key);
      }
    });
    store.connectedComponents.push(connected);
    return connected;
  };
  return store;
}
