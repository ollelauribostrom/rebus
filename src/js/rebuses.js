const rebuses = [
  {
    symbols: ['Re', '+', '🚌'],
    words: ['Rebus'],
    hint: 'You´re solving one right now'
  },
  {
    symbols: ['🏠', '+', 'pl', '+', '🐜', '+', 's'],
    words: ['Houseplants'],
    hint: `The second emoji is 'ant' not 'bug'`
  },
  {
    symbols: ['📖', '+', '🙋', '+', '📝'],
    words: ['Readme', 'file'],
    hint: 'The default markdown file of every GitHub repo'
  },
  {
    symbols: ['🚗', '+', 'a', '+', '🚐'],
    words: ['Caravan'],
    hint: 'The trailer you take when you go camping'
  },
  {
    symbols: ['⭐', '+', '🐠'],
    words: ['Starfish'],
    hint: 'Say the two emojis out loud'
  },
  {
    symbols: ['💡', '+', '🏠'],
    words: ['Lighthouse'],
    hint: 'A tower with a light to guide ships at sea'
  },
  {
    symbols: ['🌲', '+', '🍎'],
    words: ['Pineapple'],
    hint: 'Who lives in a ____ under the sea? SPONGEBOB SQUAREPANTS!'
  },
  {
    symbols: ['🥚', '+', '🌱'],
    words: ['Eggplant'],
    hint: '🍆'
  },
  {
    symbols: ['🔥', '+', '🦊'],
    words: ['Firefox'],
    hint: 'One of the most popular web browsers'
  },
  {
    symbols: ['💊', '+', 'ow'],
    words: ['Pillow'],
    hint: 'Soft item to rest your head when you sleep'
  },
  {
    symbols: ['🖊️', '+', 'd', '+', '🐜'],
    words: ['Pendant'],
    hint: `The second emoji is 'ant' not 'bug'`
  },
  {
    symbols: ['🌡️', '+', '🔑'],
    words: ['Hotkey']
  },
  {
    symbols: ['🌞', '+', '🛀'],
    words: ['Sunbath']
  },
  {
    symbols: ['🚗', '🚙', '🚗'],
    words: ['Cars'],
    hint: `They're always on the road`
  },
  {
    symbols: ['👦🏻', '+', '⚡', '+', '👓', '=', '✨'],
    words: ['Harry', 'Potter'],
    hint: 'Hogwarts'
  },
  {
    symbols: ['H', '=', 'C', '👒'],
    words: ['Cat'],
    hint: `It's raining ____ and dogs`
  },
  {
    symbols: ['FR', '=', 'D', '🐸'],
    words: ['Dog'],
    hint: `Man's best friend`
  },
  {
    symbols: ['📦', 'B', '=', 'F'],
    words: ['Fox'],
    hint: `Say the emoji out loud and replace the 'B' with 'F'`
  },
  {
    symbols: ['G', '+', '❤️'],
    words: ['Glove'],
    hint: `It covers your hand when it's cold outside`
  },
  {
    symbols: ['🍏', '-', '🐒', '+', '🐜'],
    words: ['Plant']
  },
  {
    symbols: ['🐝', '🍁'],
    words: ['Belief'],
    hint: `Hmm, there's a bee and a leaf. Say both out loud`
  },
  {
    symbols: ['🐱', '+', 'as', '+', '🏆', 'y̶', '+', 'e'],
    words: ['Catastrophe'],
    hint: 'A really really bad occurence'
  },
  {
    symbols: ['🍑', '🍑', '+', 'i', 'n', 'a', 't', 'e'],
    words: ['Assassinate'],
    hint: `What's another word for your butt? That's what the emojis mean`
  },
  {
    symbols: ['🍌', '+', '🍞'],
    words: ['Bananabread']
  },
  {
    symbols: ['🔑', '🐗', '+', 'd'],
    words: ['Keyboard'],
    hint: `You're typing with one`
  },
  {
    symbols: ['💵', '+', 'ew'],
    words: ['cashew'],
    hint: 'A type of nut'
  },
  {
    symbols: ['🤘🏻', '+', '⭐️'],
    words: ['rockstar'],
    hint: 'Mick Jagger is a what?'
  },
  {
    symbols: ['Tu', '+', '👄'],
    words: ['Tulip']
  },
  {
    symbols: ['🔥', '+', '🐶'],
    words: ['Hotdog'],
    hint: 'A close relative of burgers'
  },
  {
    symbols: ['🌙', '+', '☀'],
    words: ['Moonshine']
  },
  {
    symbols: ['💻', '+', '🌊'],
    words: ['Digital', 'Ocean'],
    hint: 'Similar to AWS and Microsoft Azure'
  },
  {
    symbols: ['🔥', '+', '🌬️', '+', '🎈'],
    words: ['Hot', 'Air', 'Balloon']
  },
  {
    symbols: ['💵', '+', '⚽'],
    words: ['Moneyball']
  },
  {
    symbols: ['🐂', '+', '👀'],
    words: ['Bullseye'],
    hint: 'The center/smallest part of a target'
  },
  {
    symbols: ['🦇', '+', 'tery'],
    words: ['Battery'],
    hint: 'All your devices contain one'
  },
  {
    symbols: ['🔒', '+', 'smith'],
    words: ['Locksmith']
  },
  {
    symbols: ['👞', '+', 'horn'],
    words: ['Shoehorn']
  },
  {
    symbols: ['🖊', '+', 't', '+', '🏠'],
    words: ['Penthouse']
  },
  {
    symbols: ['🐱', '+', 'e', '+', '💍'],
    words: ['Catering'],
    hint: 'Spell out the emojis then pronounce the word'
  },
  {
    symbols: ['2', '+', 'Na️', '+', '🐟'],
    words: ['tunafish']
  },
  {
    symbols: ['🔥', '+', 'man'],
    words: ['Fireman']
  },
  {
    symbols: ['💡', '+', 'er'],
    words: ['Lighter'],
    hint: 'Used to ignite cigarettes'
  },
  {
    symbols: ['🌙', '+', '💡'],
    words: ['Moonlight']
  },
  {
    symbols: ['f', '+', '🌬️', '+', 'y'],
    words: ['Fairy'],
    hint: 'The tooth ____ changes lost teeth to money'
  },
  {
    symbols: ['🔨', '+', '⏰'],
    words: ['Hammer', 'time'],
    hint: `MC Hammer's popular one-liner`
  },
  {
    symbols: ['👻', '+', '🚂'],
    words: ['Ghost', 'Train']
  },
  {
    symbols: ['🍦', '+', '🍦', '+', '👶'],
    words: ['Ice', 'Ice', 'Baby']
  },
  {
    symbols: ['🚗', '+', 'D', '+', '👹', '+', '🐏'],
    words: ['cardiogram'],
    hint: `Graph showing your heart's activity over a period of time`
  },
  {
    symbols: ['🌎', '+', 'wide', '+', '🕸️'],
    words: ['world', 'wide', 'web'],
    hint: `What does the 'www' in websites stand for?`
  },
  {
    symbols: ['🐀', '+', 'AT', '+', '2️', '+', 'E'],
    words: ['ratatouille'],
    hint: 'A rat chef'
  },
  {
    symbols: ['📖', '+', '🔑', '+', 'ping'],
    words: ['bookkeeping'],
    hint: 'Maintaining financial records of a business'
  },
  {
    symbols: ['sc', '+', '🦍', '+', '🐐'],
    words: ['scapegoat'],
    hint: 'Just say it out loud'
  },
  {
    symbols: ['🐜', '+', '⛵️', '+', 'tica'],
    words: ['antarctica'],
    hint: 'The southernmost continent on the planet'
  },
  {
    symbols: ['🌧', '+', '🏹'],
    words: ['rainbow'],
    hint: '🌈'
  },
  {
    symbols: ['❄️', '+', '⚾️'],
    words: ['snowball']
  },
  {
    symbols: ['❄️', '+', 'man'],
    words: ['snowman'],
    hint: '⛄'
  },
  {
    symbols: ['🐏', '📃'],
    words: ['rampage']
  },
  {
    symbols: ['✝️', '+', '🏹'],
    words: ['cross', 'bow'],
    hint: 'A sophisticated version of the bow and arrow'
  },
  {
    symbols: ['🐈', '+', '🥊'],
    words: ['cat', 'fight'],
    hint: 'AKA girl fight'
  },
  {
    symbols: ['👣', '+', '📝'],
    words: ['footnote'],
    hint: 'Placed at the bottom of a page to describe something referenced in the page'
  },
  {
    symbols: ['🐝', '+', 'r'],
    words: ['Beer'],
    hint: '🍺'
  },
  {
    symbols: ['🎉', '-', 'Y', '+', '⏰', '+', 'Job'],
    words: ['part', 'time', 'job'],
    hint: `If 🎉 means 'party', remove the 'y'`
  },
  {
    symbols: ['🌲', '+', '🏠'],
    words: ['treehouse'],
    hint: 'A house built in a tree'
  },
  {
    symbols: ['✔', '+', 'list'],
    words: ['checklist']
  },
  {
    symbols: ['💧', '+', '📦'],
    words: ['Dropbox'],
    hint: 'A popular cloud storage company; say the emojis out loud'
  },
  {
    symbols: ['💡', '+', '⚔️'],
    words: ['Lightsaber'],
    hint: 'The Star Wars weapon for hand-to-hand combat'
  },
  {
    symbols: ['☕️'],
    words: ['Java'],
    hint: 'A highly popular programming language owned by Oracle'
  },
  {
    symbols: ['⎈'],
    words: ['Helm']
  },
  {
    symbols: ['🌍', '+', 'form'],
    words: ['Terraform']
  },
  {
    symbols: ['Uni', '+', '🌽'],
    words: ['unicorn'],
    hint: 'A billion-dollar startup company'
  },
  {
    symbols: ['🆘', '+', 'desk'],
    words: ['helpdesk']
  },
  {
    symbols: ['🐱', '+', 'er', '+', '🗼'],
    words: ['caterpillar'],
    hint: 'A tiny insect with many feet'
  },
  {
    symbols: ['🧢', '+', 'tain'],
    words: ['captain'],
    hint: '_____ America 💪'
  },
  {
    symbols: ['🦊', '+', 'hole'],
    words: ['foxhole']
  },
  {
    symbols: ['🐦', '+', 's-', '+', '👁️'],
    words: ["bird's-eye"]
  },
  {
    symbols: ['🐴', '+', '🔙'],
    words: ['horseback']
  },
  {
    symbols: ['🎼', '+', '🍴'],
    words: ['pitchfork']
  },
  {
    symbols: ['🌊', '+', '🐴'],
    words: ['seahorse']
  },
  {
    symbols: ['🔙', '+', '💀'],
    words: ['backbone'],
    hint: 'Another name for your spine'
  },
  {
    symbols: ['💥', '+', '🌽'],
    words: ['popcorn'],
    hint: 'You watch movies with ___ and a drink'
  },
  {
    symbols: ['🔥', '+', '🏠'],
    words: ['firehouse']
  },
  {
    symbols: ['🚗', '+', 'go'],
    words: ['cargo'],
    hint: 'Items carried by ships and planes'
  },
  {
    symbols: ['sm', '+', '🎨'],
    words: ['smart'],
    hint: `A synonym for 'intelligent'`
  },
  {
    symbols: ['😢', '+', '👶'],
    words: ['crybaby'],
    hint: 'Someone who complains a lot'
  },
  {
    symbols: ['📻', '+', '🙂'],
    words: ['radiohead']
  },
  {
    symbols: ['🗃', '+', 'elder', '+', '🐜'],
    words: ['Box', 'Elder', 'Bug']
  },
  {
    symbols: ['L', '+', '🏹'],
    words: ['Elbow'],
    hint: 'The joint in your arm below your shoulder'
  },
  {
    symbols: ['📅', '+', 'et', '+', '🌍'],
    words: ['Planet', 'Earth']
  },
  {
    symbols: ['🔨', '+', 'head', '+', '🦈'],
    words: ['hammerhead', 'shark'],
    hint: 'A type of shark'
  },
  {
    symbols: ['👩‍', '+', '💍'],
    words: ['red', 'herring']
  },
  {
    symbols: ['Black', '+', '🎩'],
    words: ['Blackhat'],
    hint: 'A type of hacker'
  },
  {
    symbols: ['👁', '+', '💧'],
    words: ['Eyedrop'],
    hint: 'Used to treat irritated/inflammed eyes'
  },
  {
    symbols: ['🧠', '+', '🌪', '+', 'ing'],
    words: ['Brainstorming'],
    hint: 'Thinking about ideas'
  },
  {
    symbols: ['👞', '+', '📦'],
    words: ['Shoebox'],
    hunt: 'A box you get shoes in'
  },
  {
    symbols: ['💵', '+', '🛍️'],
    words: ['Money', 'Bag']
  },
  {
    symbols: ['💥', '+', 'py'],
    words: ['Poppy'],
    hint: `From Game of Thrones: 'Milk of the ____'`
  },
  {
    symbols: ['🍏', '+', '🥧'],
    words: ['Applepie']
  },
  {
    symbols: ['🍯', '+', '🌗'],
    words: ['Honeymoon'],
    hint: 'Comes right after a wedding'
  },
  {
    symbols: ['🕷', '+', '🧔🏻'],
    words: ['Spiderman'],
    hint: `Uncle Ben's superhero nephew`
  },
  {
    symbols: ['🐲', '+', '🍑'],
    words: ['Dragonfruit']
  },
  {
    symbols: ['🔫-s', '+', 'and', '+', '🥀-s'],
    words: ['Gunsandroses'],
    hint: 'A popular rock band'
  },
  {
    symbols: ['👦🏼', '+', '📔'],
    words: ['Facebook'],
    hint: `Mark Zuckerberg's company`
  },
  {
    symbols: ['T', '+', '🏃', '+', '🐱', '+', 'e'],
    words: ['Truncate']
  },
  {
    symbols: ['🐶', '+', '🐱', '+', 'cher'],
    words: ['Dogcatcher']
  },
  {
    symbols: ['🌍', '+', '⭐️'],
    words: ['Earthstar']
  },
  {
    symbols: ['👱', '+', '👨‍🎓', '+', '⛵️-s'],
    words: ['Headmasterships']
  },
  {
    symbols: ['🅰️', '+', '💣', '+', 'in', '+', '🅰️', '+', '🐂'],
    words: ['Abominable'],
    hint: 'The ___ Snowman'
  },
  {
    symbols: ['👁', '+', '💖', '+', 'YOU'],
    words: ['I', 'Love', 'You'],
    hint: '3 words, 8 letters'
  },
  {
    symbols: ['🐴', '+', '👞'],
    words: ['Horse', 'Shoe'],
    hint: 'Read the emojis out loud'
  },
  {
    symbols: ['🌍', '+', '☕'],
    words: ['World', 'Cup'],
    hint: 'A football/soccer tournament played every 4 years and organized by FIFA'
  },
  {
    symbols: ['🔁', '+', '🔙'],
    words: ['Loopback']
  },
  {
    symbols: ['🔙', '+', '🏁'],
    words: ['Backend'],
    hint: `The opposite of frontend`
  },
  {
    symbols: ['🌃', '+', '🐎'],
    words: ['Nightmare'],
    hint: 'A bad dream'
  },
  {
    symbols: ['🔙', '+', '🚪'],
    words: ['Backdoor']
  },
  {
    symbols: ['🍀', '+', '🌌‍', '+', '🚶'],
    words: ['Luke Skywalker'],
    hint: 'Star Wars: the other Skywalker (not Anakin)'
  },
  {
    symbols: ['👋', '+', '🌍'],
    words: ['Hello', 'World'],
    hint: 'The popular output of many introductory computer programming tutorials'
  },
  {
    symbols: ['🌽', '+', 'u', '+', '👮', '+', 'ia'],
    words: ['Cornucopia'],
    hint: 'The horn of plenty'
  },
  {
    symbols: ['💨', '+', '🦉', '-', 'L'],
    words: ['window'],
    hint: `The competitor of Apple, without the 's' at the end`
  },
  {
    symbols: ['❌', '+', '🔤'],
    words: ['crossword'],
    hint: 'A type of puzzle'
  },
  {
    symbols: ['S', '+', '🦀', '+', 'BLE'],
    words: ['scrabble'],
    hint: 'Words With Friends is based on this game'
  },
  {
    symbols: ['🦇', '+', '👨'],
    words: ['Batman'],
    hint: 'Bruce Wayne'
  },
  {
    symbols: ['☀️', '+', '👓'],
    words: ['Sunglasses']
  },
  {
    symbols: ['🐄', '+', '👦'],
    words: ['Cowboy']
  },
  {
    symbols: ['🖊️', '+', '👬'],
    words: ['Penpals'],
    hint: 'Friends who write letters to themselves'
  },
  {
    symbols: ['⭐', '+', '💰'],
    words: ['Starbucks'],
    hint: 'A popular American coffee company based in Seattle, Washington'
  },
  {
    symbols: ['📲', '+', '🅰️', '+', '🐀'],
    words: ['Apparat']
  },
  {
    symbols: ['☀️', '+', '🌼'],
    words: ['Sunflower']
  },
  {
    symbols: ['⛓️', '+', '🔁', '+', '🎬'],
    words: ['Chainreaction']
  },
  {
    symbols: ['🗑️', '+', '🏀'],
    words: ['Basketball'],
    hint: `Michael Jordan's primary sport`
  },
  {
    symbols: ['🌎', '+', '🐛'],
    words: ['Earthworm']
  },
  {
    symbols: ['🌬', '+', '🐟'],
    words: ['Blowfish']
  },
  {
    symbols: ['🔑', '+', '🕳'],
    words: ['Keyhole']
  },
  {
    symbols: ['💔', '+', '🕐'],
    words: ['Breaktime']
  },
  {
    symbols: ['🐊', '+', '😭'],
    words: ['Crocodile', 'tears']
  },
  {
    symbols: ['💋', '☠'],
    words: ['Kiss', 'of', 'death']
  },
  {
    symbols: ['👨', '💀', '⛵'],
    words: ['Pirate'],
    hint: 'Criminals who ransacked ships in the 1700s-1900s'
  },
  {
    symbols: ['🚪', '+', '🔔'],
    words: ['Doorbell'],
    hint: `Used to alert the occupant's of a house that somebody is at their door`
  },
  {
    symbols: ['🌙', '+', '🎂'],
    words: ['Mooncake']
  },
  {
    symbols: ['🐎', '+', '👟'],
    words: ['Horseshoe']
  },
  {
    symbols: ['⌚', '+', 'men'],
    words: ['Watchman']
  },
  {
    symbols: ['✋', '+', '📗'],
    words: ['Handbook'],
    hint: 'An instructional manual'
  },
  {
    symbols: ['🛤️', '+', 'men'],
    words: ['Railwayman']
  },
  {
    symbols: ['🐱', '+', '🐠'],
    words: ['Catfish'],
    hint: 'A person who fakes his/her identity online'
  },
  {
    symbols: ['👂', '+', '💍'],
    words: ['Earring'],
    hint: 'Say the emojis out loud'
  },
  {
    symbols: ['10', '+', '🐜'],
    words: ['Tenant'],
    hint: 'A person renting out an apartment'
  },
  {
    symbols: ['⛵️', '+', 'ment'],
    words: ['Shipment']
  },
  {
    symbols: ['2', '+', '🐝', '+', 'or', '+', 'not', '+', '2', '+', '🐝'],
    words: ['To Be Or Not To Be']
  },
  {
    symbols: ['⬇', '➡', '+', '🤮'],
    words: ['downright disgusting']
  },
  {
    symbols: ['🔥', '+', '🤼', '+', 'er'],
    words: ['firefighter'],
    hint: 'Someone trained to combat fires and rescue people trapped by fires'
  },
  {
    symbols: ['1', '+', 'ce', '+', 'n', '+', 'a', '+', '🔵', '+', '🌛'],
    words: ['Once', 'in', 'a', 'Blue Moon'],
    hint: 'Phrase used when something only happens once in a while'
  },
  {
    symbols: ['🐱', '+', '🌲'],
    words: ['catalog']
  },
  {
    symbols: ['💋', '+', 'able'],
    words: ['kissable']
  },
  {
    symbols: ['😄', '+', '🎂🎁'],
    words: ['Happy', 'Birthday'],
    hint: '____ to you!'
  },
  {
    symbols: ['🔥', '+', '⚽'],
    words: ['Fireball'],
    hint: 'The sun is a big ________'
  },
  {
    symbols: ['💎', '+', '⃝'],
    words: ['Diamond', 'ring'],
    hint: ['Things you exchange in weddings.']
  },
  {
    symbols: ['Black', '+', '🕳'],
    words: ['Black', 'hole'],
    hint: ['Final form of a star.']
  },
  {
    symbols: ['🐼', '-', 'DA', '+', '🍰'],
    words: ['Pancake'],
    hint: 'A delicious breakfast'
  },
  {
    symbols: ['✝', '+', '🏹'],
    words: ['Crossbow']
  },
  {
    symbols: ['D', '+', '💡'],
    words: ['delight']
  },
  {
    symbols: ['🌲', '+', 'T'],
    words: ['treaty'],
    hint: 'an agreement between countries'
  },
  {
    symbols: ['🔑', '+', '⛓️'],
    words: ['keychain'],
    hint: 'Holds on to one or more keys'
  }
];

export function isRebusAnswered(id) {
  const answeredRebuses = window.localStorage.getItem('answeredRebuses');
  return !!answeredRebuses && JSON.parse(answeredRebuses).includes(id);
}

export function markRebusAsAnswered(id) {
  const answeredRebuses = window.localStorage.getItem('answeredRebuses');
  if (!answeredRebuses) {
    window.localStorage.setItem('answeredRebuses', JSON.stringify([id]));
  } else {
    window.localStorage.setItem(
      'answeredRebuses',
      JSON.stringify([...JSON.parse(answeredRebuses), id])
    );
  }
}

export function getRebuses() {
  return rebuses.map((rebus, index) => {
    const id = index + 1;
    const isAnswered = isRebusAnswered(id);
    const chars = rebus.words.join('');
    return {
      id,
      ...rebus,
      input: isAnswered ? [...chars] : [...Array(chars.length)],
      isAnswered
    };
  });
}
