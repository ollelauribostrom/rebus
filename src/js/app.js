import * as Sentry from '@sentry/browser';
import { render } from './mini';
import { App } from './components/app';
import { Logo } from './components/logo';
import { GithubCorner } from './components/github-corner';
import { ChangeButton } from './components/change-button';
import { Rebus } from './components/rebus';

import { actions } from './store';
import '../css/main.css';

export function registerListeners() {
  document.addEventListener('keyup', event => {
    const key = event.key || event.keyCode; // For older browser support
    if (key === 'ArrowRight' || key === 39) {
      actions.next();
    }
    if (key === 'ArrowLeft' || key === 37) {
      actions.prev();
    }
    if (key === 'Enter' || key === 13) {
      actions.shake();
    }
  });
}

export function setCurrentFromURL() {
  const params = new URLSearchParams(window.location.search);
  const id = Number(params.get('rebus'));
  actions.setCurrent(id);
}

export function init() {
  try {
    return render(
      App(
        Logo(),
        GithubCorner({ url: 'https://github.com/ollelauribostrom/rebus' }),
        ChangeButton({
          className: 'change-button--prev',
          onClick: () => actions.prev()
        }),
        Rebus({
          charInput: (input, wordIndex, charIndex) => {
            const confettiCanon = document.querySelector('.confetti-canon');
            actions.setInput(input, wordIndex, charIndex);
            actions.check(confettiCanon);
          }
        }),
        ChangeButton({
          className: 'change-button--next',
          onClick: () => actions.next()
        })
      ),
      document.querySelector('.root')
    );
  } catch (err) {
    return Sentry.captureException(err);
  }
}

if (!global || !global.isTestRun) {
  Sentry.init({ dsn: 'https://8f025bee12e84d9b8a16e9c3b9155ce8@sentry.io/1300214' });
  init();
  registerListeners();
  setCurrentFromURL();
}
