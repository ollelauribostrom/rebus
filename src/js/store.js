import { confetti } from 'dom-confetti';
import { createStore } from './mini';
import { getRebuses, markRebusAsAnswered } from './rebuses';

export const actionsCreators = {
  next: ({ current, rebuses }) => ({
    current: current < rebuses.length - 1 ? current + 1 : 0,
    animation: 'flip-vertical-right',
    incorrectAnswerCount: 0
  }),
  prev: ({ current, rebuses }) => ({
    current: current > 0 ? current - 1 : rebuses.length - 1,
    animation: 'flip-vertical-left',
    incorrectAnswerCount: 0
  }),
  setCurrent: ({ rebuses }, id) => {
    const index = rebuses.findIndex(rebus => rebus.id === id);
    if (index > 0) {
      return { current: index };
    }
    return {};
  },
  setInput: ({ current, rebuses }, input, wordIndex, charIndex) => {
    const rebus = rebuses[current];
    const previousWords = rebus.words.slice(0, wordIndex).join('');
    const index = wordIndex > 0 ? previousWords.length + charIndex : charIndex;
    const updatedRebuses = [...rebuses];
    updatedRebuses[current].input[index] = input;
    return { updatedRebuses };
  },
  check: ({ current, rebuses, incorrectAnswerCount }, confettiCanon) => {
    const rebus = rebuses[current];
    const input = rebus.input.join('').toUpperCase();
    const answer = rebus.words.join('').toUpperCase();
    if (input.length !== answer.length) {
      return {};
    }
    if (input === answer) {
      markRebusAsAnswered(rebus.id);
      confetti(confettiCanon);
      const updatedRebuses = [...rebuses];
      updatedRebuses[current].isAnswered = true;
      return { updatedRebuses, animation: 'none', incorrectAnswerCount: 0 };
    }
    return { incorrectAnswerCount: incorrectAnswerCount + 1 };
  },
  shake: ({ current, rebuses }) => {
    const rebus = rebuses[current];
    const isAnswered = rebus.words.join('').toUpperCase() === rebus.input.join('').toUpperCase();
    if (!isAnswered) {
      return { animation: 'shake' };
    }
    return {};
  }
};

export const initialState = {
  current: 0,
  animation: 'none',
  rebuses: getRebuses(),
  incorrectAnswerCount: 0
};

export const { connect, actions } = createStore(initialState, actionsCreators);
