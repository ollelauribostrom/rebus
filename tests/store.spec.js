import * as confetti from 'dom-confetti';
import { actionsCreators, initialState } from '../src/js/store';

jest.mock('../src/js/rebuses', () => ({
  getRebuses: () => [
    {
      id: 1,
      symbols: ['😍', '👍', '😍'],
      words: ['one', 'two'],
      input: [...Array(6)],
      isAnswered: false
    },
    {
      id: 2,
      symbols: ['😍', '👍', '😍'],
      words: ['three', 'four'],
      input: [...Array(9)],
      isAnswered: false
    }
  ],
  markRebusAsAnswered: jest.fn()
}));

describe('Tests for store', () => {
  it('should have the correct initial state', () => {
    expect(initialState).toMatchSnapshot();
  });
  it('should handle action: next', () => {
    const state = { ...initialState };
    const newState = Object.assign({}, state, actionsCreators.next(state));
    const finalState = Object.assign({}, newState, actionsCreators.next(newState));
    expect(newState.current).toEqual(1);
    expect(newState.animation).toEqual('flip-vertical-right');
    expect(finalState.current).toEqual(0);
  });
  it('should handle action: prev', () => {
    const state = { ...initialState };
    const newState = Object.assign({}, state, actionsCreators.prev(state));
    const finalState = Object.assign({}, newState, actionsCreators.prev(newState));
    expect(newState.current).toEqual(1);
    expect(newState.animation).toEqual('flip-vertical-left');
    expect(finalState.current).toEqual(0);
  });
  it('should handle action: setInput', () => {
    const state = { ...initialState };
    const newState = Object.assign({}, state, actionsCreators.setInput(state, 'a', 0, 0));
    const finalState = Object.assign({}, newState, actionsCreators.setInput(newState, 'b', 1, 0));
    expect(finalState.animation).toEqual('none');
    expect(finalState.rebuses[0].input).toEqual([
      'a',
      undefined,
      undefined,
      'b',
      undefined,
      undefined
    ]);
  });
  it('should handle action: check', () => {
    const confettiCanon = document.createElement('div');
    const confettiSpy = jest.spyOn(confetti, 'confetti');
    const state = { ...initialState };
    const newState = Object.assign({}, state, actionsCreators.check(state, confettiCanon));
    expect(newState.rebuses[0].isAnswered).toEqual(false);
    newState.rebuses[0].input = ['O', 'n', 'E', 'T', 'w', 'O'];
    const finalState = Object.assign({}, newState, actionsCreators.check(newState, confettiCanon));
    expect(finalState.rebuses[0].isAnswered).toEqual(true);
    expect(confettiSpy).toHaveBeenCalledTimes(1);
    expect(confettiSpy).toHaveBeenCalledWith(confettiCanon);
  });
  it('should handle action: setCurrent', () => {
    const state = { ...initialState };
    const newState = Object.assign({}, state, actionsCreators.setCurrent(state, 2));
    const finalState = Object.assign({}, newState, actionsCreators.setCurrent(newState, 9999));
    expect(finalState.current).toEqual(1);
  });
});
