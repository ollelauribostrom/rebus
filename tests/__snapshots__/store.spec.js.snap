// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for store should have the correct initial state 1`] = `
Object {
  "animation": "none",
  "current": 0,
  "rebuses": Array [
    Object {
      "id": 1,
      "input": Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "isAnswered": false,
      "symbols": Array [
        "😍",
        "👍",
        "😍",
      ],
      "words": Array [
        "one",
        "two",
      ],
    },
    Object {
      "id": 2,
      "input": Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "isAnswered": false,
      "symbols": Array [
        "😍",
        "👍",
        "😍",
      ],
      "words": Array [
        "three",
        "four",
      ],
    },
  ],
}
`;
